### ISSUE TITLE
Can't change igenomes_base in a custom config file

### ISSUE THREAD

**Issue by ewels** [nf-core/tools#528]
### Description of bug

The iGenomes config files use `params.igenomes_base` to make it possible to use a local copy of iGenomes:

https://github.com/nf-core/tools/blob/9e33e7f02758964a1a09436af1eb6afe0e596e4c/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/conf/igenomes.config#L14

This works great when using a configuration _profile_ - the config profiles are loaded before the iGenomes config file is imported _(note line numbers)_:

https://github.com/nf-core/tools/blob/9e33e7f02758964a1a09436af1eb6afe0e596e4c/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/nextflow.config#L51-L56

https://github.com/nf-core/tools/blob/9e33e7f02758964a1a09436af1eb6afe0e596e4c/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/nextflow.config#L76-L79

However, if you use a custom configuration file, this is loaded after the pipeline `nextflow.config`. This means that `igenomes.config` has already been imported and parsed, so changing `params.igenomes_base` has no effect and the pipeline tries to use the s3 locations.

@adrlar had a nice worked example of this for nf-core/sarek [on slack](https://nfcore.slack.com/archives/CGFUX04HZ/p1581958088345100):

> If I run:
> ```bash
> nextflow -c icr.config config /mnt/scratch/DDS/SDDS/BIOINFCO/SHARED/software/sarek-2.5.2
> ```
> with the config specifying only:
> ```nextflow
> params {
>    igenomes_base = '/mnt/scratch/DDS/SDDS/BIOINFCO/SHARED/genomes/igenomes'
>    // ...
> ```
> I get:
> ```nextflow
> params {
>    // ...
>    igenomes_base = '/mnt/scratch/DDS/SDDS/BIOINFCO/SHARED/genomes/igenomes'
>    // ...
>    genomes {
>       GRCh37 {
>          acLoci = 's3://ngi-igenomes/igenomes//Homo_sapiens/GATK/GRCh37/Annotation/ASCAT/1000G_phase3_20130502_SNP_maf0.3.loci'
>          acLociGC = 's3://ngi-igenomes/igenomes//Homo_sapiens/GATK/GRCh37/Annotation/ASCAT/1000G_phase3_20130502_SNP_maf0.3.loci.gc'
>          bwaIndex = 's3://ngi-igenomes/igenomes//Homo_sapiens/GATK/GRCh37/Sequence/BWAIndex/human_g1k_v37_decoy.fasta.{amb,ann,bwt,pac,sa}'
> ```

### Proposed fix

I think it should be possible to fix this by deferring the execution of the iGenomes params paths, by surrounding them in squiggly brackets (`{}`).

However, I have a feeling that we used to do this and removed them for some reason. I just can't remember what that was.

Any issues with adding these back? Would be great if someone has some time to test this 👍

**Comment by ibwoo** [comment]
For anybody wanting a quick workaround, `--igenomes_base /path/to/igenomes/` works for me.

**Comment by adrlar** [comment]
The `--igenomes_base` command line param does indeed work to set the base before any config file gets a chance.

**Comment by ewels** [comment]
Ok, that's good - at least the command line option takes preference and is parsed in time.

I definitely have deja-vu with this one, I'm sure we've dealt with it before. Can anyone remember why we aren't using squiggly brackets @nf-core/core ?

**Comment by apeltzer** [comment]
Can't remember unfortunately, sorry :-(

**Comment by drpatelh** [comment]
This looks like the oldest commit for that file in the `nf-core/tools` repo? No `{{ }}`'s 😕 
https://github.com/nf-core/tools/blob/c2d6d4a81f15428a2cfdcce3d0140c3df98c8f99/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/conf/igenomes.config

**Comment by ewels** [comment]
Challenge accepted! Predecessor was https://github.com/nf-core/cookiecutter (which used to be called `SciLifeLab/NGI-NFcookiecutter`).

And yes, it seems that the initial commit for that repo also had no squiggly brackets: https://github.com/nf-core/cookiecutter/blob/f871965a6fd038c82ecb0e841531dfd003b2cecf/conf/igenomes.config

But before _that_ repository existed, it was developed as part of SciLifeLab/NGI-RNAseq. And **finally** I find my squiggly brackets: https://github.com/SciLifeLab/NGI-RNAseq/commit/483a3bef2787af267052daadb8f2e1d108bd0cc8 (7 Sep 2017).

But, weirdly I added the squiggly brackets and then immediately took them away again within the same pull-request: https://github.com/SciLifeLab/NGI-RNAseq/pull/152

I wonder if there was some discussion / comments that went missing from this PR or something. @Hammarn - is your memory any better than mine??

Maybe we should just try adding them in and seeing what happens 😅

**Comment by Hammarn** [comment]
@ewels it usually is, at least I tend to remember to have lunch ;) 
Eh, all I remember from this that it was a lot of changing back and forth to try and get things to work properly. 🤷‍♂

**Comment by lconde-ucl** [comment]
Hi, I tried this in a pipeline and it does not work for me.

Specifically, this is the original igenomes.config:
```
[regmond@login12 nf-core_rnaseq]$ more ~/.nextflow/assets/nf-core/rnaseq/conf/igenomes.config     
/*
 * -------------------------------------------------
 *  Nextflow config file for iGenomes paths
 * -------------------------------------------------
 * Defines reference genomes, using iGenome paths
 * Can be used by any config that customises the base
 * path using $params.igenomes_base / --igenomes_base
 */

params {
  // illumina iGenomes reference file paths
  genomes {
    'GRCh37' {
      fasta       = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa"
      bwa         = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/BWAIndex/genome.fa"
      bowtie2     = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/Bowtie2Index/"
      star        = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/"
      bismark     = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/BismarkIndex/"
      gtf         = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf"
      bed12       = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed"
      mito_name   = "MT"
      macs_gsize  = "2.7e9"
      blacklist   = "${baseDir}/assets/blacklists/GRCh37-blacklist.bed"
    }
[...]
```

When I run it with my own config, it does not pick up the igenomes that I specified:
```
----------------------------------------------------
                                        ,--./,-.
        ___     __   __   __   ___     /,-._.--~'
  |\ | |__  __ /  ` /  \ |__) |__         }  {
  | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                        `._,._,'
  nf-core/rnaseq v1.4.2
----------------------------------------------------
Pipeline Release  : master
Run Name          : agitated_swirles
Reads             : fastq_files/*{1,2}_001.fastq.gz
Data Type         : Paired-End
Genome            : GRCh37
Strandedness      : None
Trimming          : 5'R1: 0 / 5'R2: 0 / 3'R1: 0 / 3'R2: 0 / NextSeq Trim: 0
Aligner           : STAR
STAR Index        : s3://ngi-igenomes/igenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/
GTF Annotation    : s3://ngi-igenomes/igenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf
BED Annotation    : s3://ngi-igenomes/igenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed
[...]
```

Then I changed the config to add {{}}'s:
```
[regmond@login12 nf-core_rnaseq]$ more ~/.nextflow/assets/nf-core/rnaseq/conf/igenomes.config     
/*
 * -------------------------------------------------
 *  Nextflow config file for iGenomes paths
 * -------------------------------------------------
 * Defines reference genomes, using iGenome paths
 * Can be used by any config that customises the base
 * path using $params.igenomes_base / --igenomes_base
 */

params {
  // illumina iGenomes reference file paths
  genomes {
    'GRCh37' {
      fasta       = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa" }
      bwa         = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/BWAIndex/genome.fa" }
      bowtie2     = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/Bowtie2Index/" }
      star        = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/" }
      bismark     = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/BismarkIndex/" }
      gtf         = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf" }
      bed12       = { "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed" }
      mito_name   = "MT"
      macs_gsize  = "2.7e9"
      blacklist   = "${baseDir}/assets/blacklists/GRCh37-blacklist.bed"
    }
```

And it does not work (groovy/java errors and empty reference paths)"
```
N E X T F L O W  ~  version 19.10.0
Launching `nf-core/rnaseq` [deadly_mendel] - revision: 3b6df9bd10 [master]
org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.nio.file.FileSystem
org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.nio.file.FileSystem
org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.nio.file.FileSystem
----------------------------------------------------
                                        ,--./,-.
        ___     __   __   __   ___     /,-._.--~'
  |\ | |__  __ /  ` /  \ |__) |__         }  {
  | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                        `._,._,'
  nf-core/rnaseq v1.4.2
----------------------------------------------------
Pipeline Release  : master
Run Name          : deadly_mendel
Reads             : fastq_files/*{1,2}_001.fastq.gz
Data Type         : Paired-End
Genome            : GRCh37
Strandedness      : None
Trimming          : 5'R1: 0 / 5'R2: 0 / 3'R1: 0 / 3'R2: 0 / NextSeq Trim: 0
Aligner           : STAR
STAR Index        : 
GTF Annotation    : 
BED Annotation    : 
Remove Ribosomal RNA: false
Biotype GTF field : gene_biotype
Save prefs        : Ref Genome: No / Trimmed FastQ: No / Alignment intermediates: No
Max Resources     : 128 GB memory, 10 cpus, 3d time per job
[...]
```

Hopefully this feedback is useful.. :?

**Comment by ewels** [comment]
@pditommaso - any ideas why we are getting the groovy error `org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.nio.file.FileSystem` when we try to use a deferred execution string in a `params` variable for use as an input file?

**Comment by pditommaso** [comment]
Ugly, do you have the complete error stack trace?

**Comment by lconde-ucl** [comment]
Attaching the .nextflow.log file.
I didn't let the run continue, I killed it as soon as I saw the header log info
[nextflow.log](https://github.com/nf-core/tools/files/4350023/nextflow.log)

**Comment by ewels** [comment]
Minimal working example:

```nextflow
#!/usr/bin/env nextflow

params.path_1 = '/dev/null'
params.path_2 = { "/dev/null" }

in_ch_1 = Channel.fromPath(params.path_1)
in_ch_2 = Channel.fromPath(params.path_2)

println("Path 1: ${params.path_1}")
println("Path 2: ${params.path_2}")
```

```console
$ nextflow run main.nf
N E X T F L O W  ~  version 20.02.0-edge
Launching `main.nf` [sick_lorenz] - revision: c70000192e
Path 1: /dev/null
Path 2:
java.lang.String cannot be cast to java.nio.file.FileSystem
```

**Comment by pditommaso** [comment]
Ah, this is very useful.

**Comment by pditommaso** [comment]
Umm, what exactly are you expecting from the following syntax? 

```
params.path_2 = { "/dev/null" }
```

(please add @pditommaso to your reply).

**Comment by ewels** [comment]
@pditommaso In this case, I would expect it to resolve to `/dev/null` as well.

The back-story here is that we have some file paths with a `params` variable in it. When this is set / overwritten in config profiles that are _before_ this variable is defined it works fine. But if you try to set it in a `-c` config file it’s too late and it has no effect. I was hoping that by surrounding the path in squiggly brackets that it would not be resolved until later in the execution so that the overwritten `params` variable would be used.

**Comment by pditommaso** [comment]
Unfortunately no. The brackets can *only* be used in the `process` configuration block because the process has the logic to resolve them later.

**Comment by ewels** [comment]
Ok, thanks @pditommaso 

I wonder if it would work if we import the `igenomes.config` file in `main.nf` instead of in `nextflow.config` 🤔 Are command-line config files parsed before nextflow gets to `main.nf`? I think maybe that they are..

This chunk of code:

https://github.com/nf-core/tools/blob/9e33e7f02758964a1a09436af1eb6afe0e596e4c/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/nextflow.config#L76-L79

**Comment by pditommaso** [comment]
Yes, config and command line args are persed before the script run. Also, I would try to avoid conditional config includes .. :confounded:

Have you taken in consideration to include igenomes config in its own profile? then it could be activated as an extra `-profile`

**Comment by ewels** [comment]
@pditommaso That could work, but it's not the conditional load which is the problem here. It's the fact that the igenomes config file is being loaded before the `-c` config file. I don't think that having iGenomes in a profile would change that.

**Comment by pditommaso** [comment]
It should because, the order in which the profiles are merged it depends on how you mention in the `-profile` argument I mean `-profile foo,bar` is different from `-profile bar,foo`. Could not address your issue?

**Comment by ewels** [comment]
@pditommaso No, because this is specifically for configs loaded with `-c`, which is always after all profiles.

**Comment by pditommaso** [comment]
I see. What is preventing to move it as a base profile as it is done for `docker`, `test`, etc.? Asking for a friend :)

https://github.com/nf-core/rnaseq/blob/master/nextflow.config#L119-L128

**Comment by ewels** [comment]
@pditommaso Same thing - we _always_ want it to load as late as possible so that the variable in the string can be changed in as many places as possible. We moved it after the profiles so that all profiles could change that string and that worked. But it's not enough for `-c` config flags. I think if we move it to `main.nf` then that might work for all cases though.. (hopefully)

**Comment by pditommaso** [comment]
Ok, I go through all the thread, I had started from the error comment. Unfortunately, this is expected. The config was designed to allow the params override from the command, not across config files. 

Maybe it should be taken in consideration to handle the igenome dataset not as a config file, but for example, as a ymal file that could be read the from the NF script. The `igenomes_base` would remain as a parameter that could be used to parametrized the yaml file. 

Currently, it could be done  with a few lines of custom code. But it would be nice to have a NF built-in support for this.

**Comment by ewels** [comment]
Sure, that could be nice! Though to be honest I am hesitant to completely replace what we currently have as generally it works so well 😅 It's super nice that you can add to this list of genomes / overwrite specific paths and generally mess around with it using the built-in nextflow config handling. It would probably be possible to maintain this functionality with a YAML approach, but it would need some thought.

@lconde-ucl - as our chief tester, do you think you could try my suggestion mentioned here? https://github.com/nf-core/tools/issues/528#issuecomment-601153463 - just pop those lines in near the top of `main.nf` and give it a try..

**Comment by lconde-ucl** [comment]
@ewels - adding these lines in main.nf doesn't work:

```
[regmond@login12 nf-core_rnaseq]$ nextflow run nf-core/rnaseq --reads 'fastq_files/*{1,2}_001.fastq.gz' --genome GRCh37 -c "/scratch/scratch/regmond/ucl.config" 
N E X T F L O W  ~  version 19.10.0
Launching `nf-core/rnaseq` [small_stonebraker] - revision: 3b6df9bd10 [master]
Unknown method `includeConfig`

 -- Check script '/home/regmond/.nextflow/assets/nf-core/rnaseq/main.nf' at line: 119 or see '.nextflow.log' file for more details

```

**Comment by pditommaso** [comment]
`includeConfig` is not available in the script context.

**Comment by drpatelh** [comment]
Yeah, I've failed with trying that in the past too. What if we were to add `igenomes.config` to `nf-core/configs` and pull it from there? 🤔 Would we be able to get around the config parsing order that way?

**Comment by ewels** [comment]
Nope, I don't think so. Every config thing that we can possibly do is always going to be before the `-c` option. I think our only option now is to build this config array in the main script using custom code.

To be completely honest, I'm not super keen on a complete rewrite for what is a fairly minor edge case. I'm more inclined to just document this and how to manually resolve it. What does everyone think?

**Comment by drpatelh** [comment]
I think its fine to leave it as it is and document it as you suggested. Worse case scenario you just have to provide `--igenomes_base` on the command-line which is is fine. If and when the custom config is uploaded to `nf-core/configs` and contains the `--igenomes_base` parameter it should work as expected anyway.

**Comment by ewels** [comment]
I guess we can add this to https://nf-co.re/usage/troubleshooting ..?

**Comment by drpatelh** [comment]
Or here: https://nf-co.re/usage/reference_genomes#illumina-igenomes ? or to both 😅

**Comment by MaxUlysse** [comment]
> I guess we can add this to https://nf-co.re/usage/troubleshooting ..?

> Or here: https://nf-co.re/usage/reference_genomes#illumina-igenomes ? or to both sweat_smile

Can we actually mention that somewhere, just spend 30 min on that issue again...

**Comment by MaxUlysse** [comment]
Actually it was not exactly this one, but a similar one with `igenomes_ignore`
